drivers/i2s/i2s_litex.c

```diff
@@ -0,0 +1,617 @@
+ * This function copies data from user buffer into fifo
+ *
+ * @param src memory from which data will be copied to fifo
+ * @param size amount of data to be copied in bytes
+ * @param sample_width width of signle sample in bits
+ * @param channels number of received channels
+ *
+ * @return N/A
+ */
+static void i2s_copy_to_fifo(uint8_t *src, size_t size, int sample_width,
+			     int channels)
+{
+	int chan_size = sample_width / 8;
+#if CONFIG_I2S_LITEX_CHANNELS_CONCATENATED
+	if (channels == 2) {
+		for (size_t i = 0; i < size / chan_size; i += 4) {
+			//using sys_write function, as fifo is not a csr,
+			//but a contignous memory space
+			sys_write32(*(src + i), I2S_TX_FIFO_ADDR);
+		}
+	} else {
+		for (size_t i = 0; i < size / chan_size; i += 2) {
+			sys_write32(*((uint16_t*)(src + i)), I2S_TX_FIFO_ADDR);
+		}
+	}
+#else
+	int max_off = chan_size - 1;
+	uint32_t data;
+	uint8_t *d_ptr = (uint8_t *)&data;
+	for (size_t i = 0; i < size / chan_size; ++i) {
+		for (int off = max_off; off >= 0; off--) {
+#if CONFIG_I2S_LITEX_DATA_BIG_ENDIAN
+				*(d_ptr + off) = *(src + i * chan_size + (max_off - off));
+#else
+				*(d_ptr + off) = *(src + i * chan_size + off);
+#endif
+
+		}
+		sys_write32(data, I2S_TX_FIFO_ADDR);
+		// if mono send every left channel
+		// right channel will be same as left
+		if (channels == 1) {
+			sys_write32(data, I2S_TX_FIFO_ADDR);
+		}
+	}
+#endif
+}
+
+/*
+ * Get data from the queue
+ */
+static int queue_get(struct ring_buf *rb, void **mem_block, size_t *size)
+{
+	unsigned int key;
+
+	key = irq_lock();
+
+	if (rb->tail == rb->head) {
+		/* Ring buffer is empty */
+		irq_unlock(key);
+		return -ENOMEM;
+	}
+	*mem_block = rb->buf[rb->tail].mem_block;
+	*size = rb->buf[rb->tail].size;
+	MODULO_INC(rb->tail, rb->len);
+
+	irq_unlock(key);
+	return 0;
+}
+
+/*
+ * Put data in the queue
+ */
+static int queue_put(struct ring_buf *rb, void *mem_block, size_t size)
+{
+	uint16_t head_next;
+	unsigned int key;
+
+	key = irq_lock();
+
+	head_next = rb->head;
+	MODULO_INC(head_next, rb->len);
+
+	if (head_next == rb->tail) {
+		/* Ring buffer is full */
+		irq_unlock(key);
+		return -ENOMEM;
+	}
+
+	rb->buf[rb->head].mem_block = mem_block;
+	rb->buf[rb->head].size = size;
+	rb->head = head_next;
+
+	irq_unlock(key);
+	return 0;
+}
+
+static int i2s_litex_initialize(struct device *dev)
+{
+	struct i2s_litex_cfg *cfg = DEV_CFG(dev);
+	struct i2s_litex_data *const dev_data = DEV_DATA(dev);
+	k_sem_init(&dev_data->rx.sem, 0, CONFIG_I2S_LITEX_RX_BLOCK_COUNT);
+	k_sem_init(&dev_data->tx.sem, CONFIG_I2S_LITEX_TX_BLOCK_COUNT - 1,
+		   CONFIG_I2S_LITEX_TX_BLOCK_COUNT);
+
+	cfg->irq_config(dev);
+	return 0;
+}
+
+static int i2s_litex_configure(struct device *dev, enum i2s_dir dir,
+			       struct i2s_config *i2s_cfg)
+{
+	struct i2s_litex_data *const dev_data = DEV_DATA(dev);
+	const struct i2s_litex_cfg *const cfg = DEV_CFG(dev);
+	struct stream *stream;
+	int channels_concatenated;
+	if (dir == I2S_DIR_RX) {
```

Comment:
Add a space between the variable definition and the body.

Reply:
added